languages/java/oso/src/main/java/com/osohq/oso/Polar.java

```diff
@@ -88,15 +92,65 @@ public void loadStr(String str) throws Exceptions.OsoException {
   }
 
   /** Query for a predicate, parsing it first. */
-  public Query query(String query) throws Exceptions.OsoException {
-    return new Query(ffiPolar.newQueryFromStr(query), host.clone());
+  public Query query(String query) throws OsoException {
+    return query(query, Map.of(), false);
+  }
+
+  /** Query for a predicate, parsing it first and optionally accepting an expression. */
+  public Query query(String query, boolean acceptExpression) throws OsoException {
+    return query(query, Map.of(), acceptExpression);
+  }
+
+  /** Query for a predicate, parsing it first and applying bindings */
+  public Query query(String query, Map<String, Object> bindings) throws Exceptions.OsoException {
+    return query(query, bindings, false);
+  }
+
+  /**
+   * Query for a predicate, parsing it first, applying bindings and optionally accepting an
+   * expression.
+   */
+  public Query query(String query, Map<String, Object> bindings, boolean acceptExpression)
+      throws Exceptions.OsoException {
+    Host new_host = host.clone();
+    new_host.setAcceptExpression(acceptExpression);
+    return new Query(ffiPolar.newQueryFromStr(query), new_host, bindings);
   }
 
   /** Query for a predicate. */
   public Query query(Predicate query) throws Exceptions.OsoException {
     Host new_host = host.clone();
     String pred = new_host.toPolarTerm(query).toString();
-    return new Query(ffiPolar.newQueryFromTerm(pred), new_host);
+    return new Query(ffiPolar.newQueryFromTerm(pred), new_host, Map.of());
+  }
+
+  /** Query for a predicate, optionally accepting expressions in the result. */
+  public Query query(Predicate query, boolean acceptExpression) throws Exceptions.OsoException {
+    return query(query, Map.of(), acceptExpression);
+  }
+
+  /**
+   * Query for a predicate, applying bindings and optionally accepting the expression type as a
+   * result.
+   *
+   * @param acceptExpression Set to true to accept an Expression as a result from the VM.
+   */
+  public Query query(Predicate query, Map<String, Object> bindings, boolean acceptExpression)
+      throws Exceptions.OsoException {
+    Host new_host = host.clone();
+    new_host.setAcceptExpression(acceptExpression);
```

Comment:
Why don't we make `setAcceptExpression` a method on the Query object that calls this? And then we don't need the multiple APIs for it?

Reply:
Unfortunately Query gets a result when it is constructed from the core, so this parameter needs to be set on Host before query construction.

This also (roughly) matches the API in other languages, but I agree it's not ideal.